From: Martin Schwidefsky <schwidefsky@de.ibm.com>
Subject: s390: report new vector facilities
Patch-mainline: v4.11-rc1
Git-commit: 2583b848cad049cf5f3f0a03af8b140668b376f3
References: FATE#324088, LTC#158828

Summary:     kernel: SIMD extensions for z14
Description: With this feature, Linux user space applications can detect
             the availability of the 'vector decimal facility' and the
             'vector enhancement facility 1' extensions of the IBM z
             vector facility. 
             These extensions are designed to improve the performance of
             analytic workloads.

Upstream-Description:

             s390: report new vector facilities

             Add hardware capability bits and feature tags to /proc/cpuinfo for
             the "Vector Packed Decimal Facility" (tag "vxd" / hwcap bit 2^12)
             and the "Vector Enhancements Facility 1" (tag "vxe" / hwcap bit 2^13).

             Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>


Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>
Acked-by: Johannes Thumshirn <jthumshirn@suse.de>
---
 arch/s390/include/asm/elf.h  |    2 ++
 arch/s390/kernel/processor.c |    2 +-
 arch/s390/kernel/setup.c     |    7 ++++++-
 3 files changed, 9 insertions(+), 2 deletions(-)

--- a/arch/s390/include/asm/elf.h
+++ b/arch/s390/include/asm/elf.h
@@ -103,6 +103,8 @@
 #define HWCAP_S390_HIGH_GPRS	512
 #define HWCAP_S390_TE		1024
 #define HWCAP_S390_VXRS		2048
+#define HWCAP_S390_VXRS_BCD	4096
+#define HWCAP_S390_VXRS_EXT	8192
 
 /* Internal bits, not exposed via elf */
 #define HWCAP_INT_SIE		1UL
--- a/arch/s390/kernel/processor.c
+++ b/arch/s390/kernel/processor.c
@@ -59,7 +59,7 @@ static int show_cpuinfo(struct seq_file
 {
 	static const char *hwcap_str[] = {
 		"esan3", "zarch", "stfle", "msa", "ldisp", "eimm", "dfp",
-		"edat", "etf3eh", "highgprs", "te", "vx"
+		"edat", "etf3eh", "highgprs", "te", "vx", "vxd", "vxe"
 	};
 	static const char * const int_hwcap_str[] = {
 		"sie"
--- a/arch/s390/kernel/setup.c
+++ b/arch/s390/kernel/setup.c
@@ -753,8 +753,13 @@ static int __init setup_hwcaps(void)
 	 * can be disabled with the "novx" parameter. Use MACHINE_HAS_VX
 	 * instead of facility bit 129.
 	 */
-	if (MACHINE_HAS_VX)
+	if (MACHINE_HAS_VX) {
 		elf_hwcap |= HWCAP_S390_VXRS;
+		if (test_facility(134))
+			elf_hwcap |= HWCAP_S390_VXRS_EXT;
+		if (test_facility(135))
+			elf_hwcap |= HWCAP_S390_VXRS_BCD;
+	}
 	get_cpu_id(&cpu_id);
 	add_device_randomness(&cpu_id, sizeof(cpu_id));
 	switch (cpu_id.machine) {
